File: packages/contracts/src/file.ts
import type { AuditAction } from "./audit";
import type { ID, ISODateTime } from "./global";

export interface FileRef {
  fileId: ID;
  name: string;
  mimeType: string;
  size: number;
  url?: string;
}

export type FileStatus = "PENDING" | "UPLOADING" | "READY" | "FAILED" | "DELETED";

export type FileWorkflowCommand =
  | "createUploadUrl"
  | "markUploadStarted"
  | "confirmUpload";

export type FileUploadLifecycleRule =
  | {
      command: "createUploadUrl";
      fromStatus?: never;
      toStatus: "PENDING";
      auditAction: Extract<AuditAction, "FILE_UPLOAD_URL_CREATED">;
    }
  | {
      command: "markUploadStarted";
      fromStatus: "PENDING";
      toStatus: "UPLOADING";
      auditAction: Extract<AuditAction, "FILE_UPLOAD_STARTED">;
    }
  | {
      command: "confirmUpload";
      fromStatus: "PENDING" | "UPLOADING";
      toStatus: "READY";
      auditAction: Extract<AuditAction, "FILE_CONFIRMED">;
    };

export interface FileObject {
  id: ID;
  ownerId: ID;
  ownerType: "USER" | "ASSIGNMENT" | "EXPORT_JOB";
  purpose: "DATASET_IMPORT" | "ANSWER_ATTACHMENT" | "EXPORT_RESULT";
  mimeType: string;
  size: number;
  storageKey: string;
  status: FileStatus;
  createdAt: ISODateTime;
  confirmedAt?: ISODateTime;
}

export interface CreateUploadUrlRequest {
  fileName: string;
  mimeType: string;
  size: number;
  purpose: "DATASET_IMPORT" | "ANSWER_ATTACHMENT" | "EXPORT_RESULT";
  ownerType: "USER" | "ASSIGNMENT" | "EXPORT_JOB";
  ownerId: ID;
}

export interface CreateUploadUrlResponse {
  file: FileObject;
  uploadUrl: string;
  headers?: Record<string, string>;
  expiresAt: ISODateTime;
}

export interface ConfirmUploadRequest {
  storageKey?: string;
  checksum?: string;
}

export interface ConfirmUploadResponse {
  file: FileObject;
}

File: apps/api/app/schemas/file.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class FileObjectResponse(BaseModel):
    id: str
    ownerId: str
    ownerType: str
    purpose: str
    mimeType: str
    size: int
    uploadedSize: int | None
    checksumSha256: str | None
    storageKey: str
    status: str
    createdAt: datetime
    confirmedAt: datetime | None

    @classmethod
    def from_orm(cls, f) -> "FileObjectResponse":
        return cls(
            id=f.id, ownerId=f.owner_id, ownerType=f.owner_type,
            purpose=f.purpose, mimeType=f.mime_type, size=f.size,
            uploadedSize=f.uploaded_size,
            checksumSha256=f.checksum_sha256,
            storageKey=f.storage_key, status=f.status,
            createdAt=f.created_at, confirmedAt=f.confirmed_at,
        )


class CreateUploadUrlRequest(BaseModel):
    fileName: str = Field(min_length=1, max_length=255)
    mimeType: str = Field(min_length=1, max_length=255)
    size: int = Field(gt=0)
    purpose: str        # DATASET_IMPORT / ANSWER_ATTACHMENT / EXPORT_RESULT
    ownerType: str      # USER / ASSIGNMENT / EXPORT_JOB
    ownerId: str


class CreateUploadUrlResponse(BaseModel):
    file: FileObjectResponse
    uploadUrl: str
    headers: dict | None = None
    expiresAt: datetime


class ConfirmUploadRequest(BaseModel):
    storageKey: str | None = Field(default=None, max_length=500)
    checksum: str | None = Field(default=None, max_length=80)


class ConfirmUploadResponse(BaseModel):
    file: FileObjectResponse
